### PR TITLE
Physics: remove *VelocityCmd at each time step

### PR DESCRIPTION
**Pull request by scpeters**
# 🎉 New feature

Part of #1926.

## Summary

This implements the suggestion from https://github.com/gazebosim/gz-sim/issues/1926#issuecomment-1472113044 to delete all `*VelocityCmd` components after each time step.

## Test it

I'm still figuring out how to test it.

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by scpeters** [comment]
cleaned up some code and improved coverage with `SetComponentData` in https://github.com/gazebosim/gz-sim/pull/2228/commits/b145c05aed9f537191e795354f3a8d124c80ed46

**Review by azeey — COMMENTED** [review]
Looks good so far!

It would be good to characterize which systems are changing behavior due to this. For example, it looks like the DiffDrive system won't change behavior because it'll constantly be sending the last twist command it received.

File: src/systems/multicopter_motor_model/MulticopterMotorModel.cc [hunk 1]
```diff
@@ -447,14 +447,6 @@ void MulticopterMotorModel::PreUpdate(const UpdateInfo &_info,
     doUpdateForcesAndMoments = false;
   }
 
-  if (!_ecm.Component<components::JointVelocityCmd>(
-      this->dataPtr->jointEntity))
-  {
-    _ecm.CreateComponent(this->dataPtr->jointEntity,
-        components::JointVelocityCmd({0}));
-    doUpdateForcesAndMoments = false;
```
[the remark this answers is not in the document]
Comment (scpeters) [reply src/systems/multicopter_motor_model/MulticopterMotorModel.cc:455]: no, the other places that variable is set to false is when a physics state / sensor reading component is unavailable that is needed for the control calculation, but `JointVelocityCmd` is an actuator command written by this system, so it's ok if it's missing

**Comment by scpeters** [comment]
> It would be good to characterize which systems are changing behavior due to this. For example, it looks like the DiffDrive system won't change behavior because it'll constantly be sending the last twist command it received.

the following systems reference a `*VelocityCmd` component:

~~~
$ grep -rlI VelocityCmd src/systems | sort 
src/systems/ackermann_steering/AckermannSteering.cc
src/systems/battery_plugin/LinearBatteryPlugin.cc
src/systems/diff_drive/DiffDrive.cc
src/systems/joint_controller/JointController.cc
src/systems/joint_position_controller/JointPositionController.cc
src/systems/mecanum_drive/MecanumDrive.cc
src/systems/multicopter_motor_model/MulticopterMotorModel.cc
src/systems/physics/Physics.cc
src/systems/thruster/Thruster.cc
src/systems/trajectory_follower/TrajectoryFollower.cc
src/systems/velocity_control/VelocityControl.cc
~~~

* The LinearBatteryPlugin only reads the content of a `JointVelocityCmd`; it does not modify them, so it's behavior will not change
* The Physics system does not set the content of any `*VelocityCmd` components; it only removes them via the changes in this PR

The following systems are already modified by this pull request to use the `SetComponentData` API, and I have checked them to verify that their behavior will not change.

~~~
src/systems/ackermann_steering/AckermannSteering.cc
src/systems/diff_drive/DiffDrive.cc
src/systems/joint_controller/JointController.cc
src/systems/joint_position_controller/JointPositionController.cc
src/systems/multicopter_motor_model/MulticopterMotorModel.cc
src/systems/thruster/Thruster.cc
src/systems/velocity_control/VelocityControl.cc
~~~

I still need to review the following two systems:

* src/systems/mecanum_drive/MecanumDrive.cc
* src/systems/trajectory_follower/TrajectoryFollower.cc

**Comment by scpeters** [comment]
> I still need to review the following two systems:
> 
> * src/systems/mecanum_drive/MecanumDrive.cc
> * src/systems/trajectory_follower/TrajectoryFollower.cc

Upon further review, I found that the `MecanumDrive` system's behavior is unchanged, though I took the opportunity to use the `SetComponentData` API in https://github.com/gazebosim/gz-sim/pull/2228/commits/7dc33fe70d47312acd45406335547fec08a19fa2. The `TrajectoryFollower` system was already removing `AngularVelocityCmd` components at every timestep that it didn't want to set them, which is now not necessary, so I removed that logic in https://github.com/gazebosim/gz-sim/pull/2228/commits/120e5fef83cbf8fdca6649ee92c30ed09fd357fc

**Comment by scpeters** [comment]
I also noticed that the MecanumDrive and TrajectoryFollower systems don't have integration tests, so I filed #2238 and #2239.

**Comment by mjcarroll** [comment]
So not something that needs to be addressed here, but I believe that adding and removing components is relatively expensive (compared to setting/unsetting them) as they can incur allocation/deallocation as well as additional bookkeeping inside the ECM.

I think in the future, we should consider having an `optional<Component>` type that remains attached to an entity, but isn't created/destroyed each iteration.

**Comment by scpeters** [comment]
> I think in the future, we should consider having an `optional<Component>` type that remains attached to an entity, but isn't created/destroyed each iteration.

That could make sense if it makes a significant performance improvement. We could define each component as `Component<std::optional<T> ...`, but I'm guessing your proposal sounds a little different than that

**Comment by mjcarroll** [comment]
> That could make sense if it makes a significant performance improvement.

I don't have a good gauge of the impact.  I'm going to assume that not allocating/bookkeeping are always going to be better than allocation/bookkeeping.  I think in this case, the trade-off would be a small amount of memory overhead, because you would be keeping the components in question (or a pointer to them).

> We could define each component as Component<std::optional<T> ..., but I'm guessing your proposal sounds a little different than that

That's basically what I'm suggesting.  We may be able to do a little bit of syntactic sugar to make an `OptionalComponent<T>`. 
 That could enable letting `Each` calls only iterate over components that are set/unset.

**Comment by mjcarroll** [comment]
as a followup on the performance side of things, the concept of "optional components" was mostly added already: https://github.com/gazebosim/gz-sim/issues/1334#issuecomment-1036570727

So that is one less thing to worry about here.

**Comment by azeey** [comment]
@scpeters what's the status of this PR?

**Comment by scpeters** [comment]
> @scpeters what's the status of this PR?

sorry I missed this notification until now. I thought it was good to go in November, but it may have gone out of date since then. I can open a separate pull request for the `SetComponentData` changes in the other systems to reduce the noise if that would help

**Comment by scpeters** [comment]
> I can open a separate pull request for the `SetComponentData` changes in the other systems to reduce the noise if that would help

https://github.com/gazebosim/gz-sim/pull/2360

**Comment by scpeters** [comment]
> > I can open a separate pull request for the `SetComponentData` changes in the other systems to reduce the noise if that would help
> 
> #2360

I merged these changes to `gz-sim8` and then forward to `main` and merged into this branch, so the diff is a bit smaller now. This is ready for review again

File: Migration.md [hunk 2]
```diff
@@ -5,6 +5,13 @@ Deprecated code produces compile-time warnings. These warning serve as
 notification to users that their code should be upgraded. The next major
 release will remove the deprecated code.
 
+## Gazebo Sim 8.x to 9.0
+
+ * **Modified**:
+   + In the Physics system, all `*VelocityCmd` components are deleted after
```
Comment (iche033) [inline Migration.md:11]: mention that previously it was setting them to zero?

**Review by iche033 — APPROVED** [review]
looks good to me

File: Migration.md (hunk 2 above)
Comment (scpeters) [reply Migration.md]: https://github.com/gazebosim/gz-sim/pull/2228/commits/f4ffd4a8721915a8c9ac90ab45006927835b7bf3

File: Migration.md [hunk 3]
```diff
@@ -5,6 +5,14 @@ Deprecated code produces compile-time warnings. These warning serve as
 notification to users that their code should be upgraded. The next major
 release will remove the deprecated code.
 
+## Gazebo Sim 8.x to 9.0
+
+ * **Modified**:
```
Comment (azeey) [inline Migration.md:10]: Thanks for adding this!

File: test/worlds/quadcopter.sdf [hunk 4]
```diff
@@ -91,7 +91,7 @@
           </geometry>
         </collision>
         <visual name="rotor_0_visual">
-          <pose>0 0 0 1.57 0 0 0</pose>
+          <pose>0 0 0 1.57 0 0</pose>
```
Comment (azeey) [inline test/worlds/quadcopter.sdf:94]: I realized last week while trying to fix another test that the quadcopter test is passing even though this SDF was invalid. I'll be making the same changes in other branches.

**Review by azeey — APPROVED** [review]

**Pull request merged**